File: src/main/java/com/trip_service/trip_service/route/controllers/VehicleController.java
package com.trip_service.trip_service.route.controllers;

import com.trip_service.trip_service.route.models.VehicleDetails;
import com.trip_service.trip_service.route.services.VehicleService;
import com.trip_service.trip_service.route.DTO.RequestBodyDTO;
import com.trip_service.trip_service.utils.DTO.SendResponse;
import com.trip_service.trip_service.utils.exceptions.GeneratedApiException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/vehicle")
public class VehicleController {

    @Autowired
    private VehicleService vehicleService;

    @PostMapping("/register")
    @PreAuthorize("hasAnyAuthority('ADMIN', 'TRIP_ADMIN', 'VEHICLE_ADMIN')")
    public SendResponse<VehicleDetails> registerVehicle(@RequestBody RequestBodyDTO.registerVehicleRequestBody requestBody) {

        try {
            return new SendResponse<>(
                    200,
                    "Vehicle Registered Successfully",
                    "",
                    vehicleService.registerVehicle(requestBody)
            );
        }
        catch (GeneratedApiException apiException){
            return new SendResponse<>(apiException.getStatusCode(), apiException.getMessage(), apiException.getErrorMsg(), null);
        }
        catch (Exception e) {
            return new SendResponse<>(500, "Internal Server Error!!", e.getMessage(), null);
        }

    }


}

File: src/main/kotlin/com/trip_service/trip_service/core/vehicles/controllers/VehicleController.kt
package com.trip_service.trip_service.core.vehicles.controllers

import com.trip_service.trip_service.core.vehicles.dto.VehiclesRequestBody
import com.trip_service.trip_service.core.vehicles.models.Vehicles
import com.trip_service.trip_service.core.vehicles.services.VehicleServices
import com.trip_service.trip_service.helpers.dto.GenerateResponse
import com.trip_service.trip_service.helpers.errors.GenerateApiException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/vehicles")
class VehicleController {

    @Autowired
    private lateinit var vehicleServices: VehicleServices;

    /*
    Create Vehicle
    */
    @PostMapping("/register")
    fun createVehicle(@RequestBody vehicle: VehiclesRequestBody.CreateVehicles): GenerateResponse<String> {

        try {

            vehicleServices.createNewVehicle(vehicle);

            return GenerateResponse(200, "Vehicle Created Successfully!", "", null);

        }
        catch (exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage,
                null
            )
        }

    }

    /*
    Get All  Vehicle Details
     */
    @PostMapping("/all")
    fun getAllVehicles(): GenerateResponse<List<Vehicles>> {

        try {

            return GenerateResponse(
                200,
                "Vehicle Retrived Successfully!!",
                "",
                vehicleServices.getAllVehicles()
            );

        }
        catch (exception: GenerateApiException) {
            return GenerateResponse(
                exception.status,
                "",
                exception.errorMessage,
                null
            );
        }

    }

}